### DISCUSSION TITLE
MPR channels not working?

### DISCUSSION THREAD

**Discussion by fmw42** [ImageMagick/ImageMagick#6358] [Digital Image Processing]
Why is this process not reversible?  Am I not getting different channels and processing the same channel? What am I missing?

Input:

![img](https://github.com/ImageMagick/ImageMagick/assets/13458218/bc7fa2a3-3c7b-4c24-bf7d-f7f0c7992df9)

```
magick img.png \
-colorspace HSV -write mpr:HSV +delete \
mpr:HSV[0] mpr:HSV[1] mpr:HSV[2] \
-set colorspace HSV -combine -colorspace sRGB \
out.png
```

![out](https://github.com/ImageMagick/ImageMagick/assets/13458218/2f6e55cd-b63d-4866-aa62-3500c64c91a3)

**Comment by fmw42** [comment]
It appears that I have to **separate** channels before saving to MPR:.  However, that loses the alpha channel, so I have to save it first and add it back.

However, there is a small black spot on the upper left red area in the output that is not in the input.

Input:

![img](https://github.com/ImageMagick/ImageMagick/assets/13458218/3074ffed-bfe8-43ce-8705-ea043ac2eda9)

```
magick img.png \
\( +clone -alpha extract -write mpr:alpha +delete \) \
-colorspace HSV -separate -write mpr:HSV -delete 0--1 \
mpr:HSV[0] mpr:HSV[1] mpr:HSV[2] \
-combine -set colorspace HSV -colorspace sRGB \
mpr:alpha -alpha off -compose copy_opacity -composite \
out.png

```

![out](https://github.com/ImageMagick/ImageMagick/assets/13458218/00f406e0-8a0a-4cb5-9761-1ceea9cd6352)


Why do I need to separate channels after changing colorspace and before saving to MPR:?  And why that small black spot?

**Comment by fmw42** [comment]
Doing the same in RGB results in a grayscale image. So that seems to indicate that -separate must be added. Without it, -write MPR: is apparently only saving one channel or some combination of channels.  

 **Is this a bug? Should it not save all channels of the image? Or does MPR: not permit channel specifiers?**


```
magick img.png \
-write mpr:RGB -delete 0--1 \
\( mpr:RGB[0] \) \
\( mpr:RGB[1] \) \
\( mpr:RGB[2] \) \
-combine -colorspace sRGB \
out.png
```

![out](https://github.com/ImageMagick/ImageMagick/assets/13458218/c70bcc67-c710-4cac-b1d5-fcb344d3071f)

**Comment by snibgo** [comment]
In this command...
```
magick img.png \
-colorspace HSV -write mpr:HSV +delete \
mpr:HSV[0] mpr:HSV[1] mpr:HSV[2] \
-set colorspace HSV -combine -colorspace sRGB \
out.png
```
... you write one image to mpr:HSV. Only one image, with three channels.

Then you read the first image `mpr:HSV[0]` which is fair enough. Then you read the second image `mpr:HSV[1]`, but there is no second image, so IM gives you the first again. Likewise the third image.

IM should really throw an error when trying to read `mpr:HSV[1]` and `mpr:HSV[2]`. But it doesn't.

If you want three grayscale images to be saved as `mpr:HSV` so you can combine them later, you need to separate them first.

When the input has three colour channel plus an alpha channel, `-separate` will make four images (unless we have "`-channel RGB`" etc first).

**Reply by fmw42** [reply]
Why is there no mpr:HSV[1]?   I saved the HSV image with 3 channels with -write mpr:HSV?  Sorry I don't get why the -separate is necessary. Is -write mpr:HSV not saving all 3 channels or is it an issue with accessing channels [X]?

**Comment by snibgo** [comment]
"`-write mpr:HSV`" is saving one image with three channels. Only one image.

The [N] suffix refers to *image* numbers, not *channel* numbers, starting from zero.

**Reply by fmw42** [reply]
Thanks. DOH! Not sure where my head was.
